app/code/Magento/SalesRule/Observer/UpdateQuoteOnDisableSalesrule.php

```diff
@@ -0,0 +1,104 @@
+<?php
+
+namespace Magento\SalesRule\Observer;
+
+use Magento\Framework\Event\Observer;
+use Magento\Framework\Event\ObserverInterface;
+use Magento\Framework\Model\ResourceModel\Iterator;
+use Magento\Quote\Api\CartRepositoryInterface;
+use Magento\Quote\Model\Quote;
+use Magento\Quote\Model\ResourceModel\Quote\Collection;
+use Magento\Quote\Model\ResourceModel\Quote\CollectionFactory;
+use Magento\SalesRule\Model\Rule;
+use Psr\Log\LoggerInterface;
+
+class UpdateQuoteOnDisableSalesrule implements ObserverInterface
+{
+    /**
+     * @var CollectionFactory
+     */
+    private $collectionFactory;
+
+    /**
+     * @var LoggerInterface
+     */
+    private $logger;
+
+    /**
+     * @var Iterator
+     */
+    private $iterator;
+
+    /**
+     * @var CartRepositoryInterface
+     */
+    private $cartRepository;
+
+    /**
+     * UpdateQuoteOnDisableSalesrule constructor.
+     * @param Iterator $iterator
+     * @param CollectionFactory $collectionFactory
+     * @param CartRepositoryInterface $cartRepository
+     */
+    public function __construct(
+        Iterator $iterator,
+        CollectionFactory $collectionFactory,
+        CartRepositoryInterface $cartRepository,
+        LoggerInterface $logger
+    ) {
+        $this->iterator = $iterator;
+        $this->collectionFactory = $collectionFactory;
+        $this->cartRepository = $cartRepository;
+        $this->logger = $logger;
+    }
+
+    /**
+     * We should ignore active rules and recalculate quotes only when rule is disabled.
+     *
+     * @param Observer $observer
+     * @return void
+     */
+    public function execute(Observer $observer): void
+    {
+        /** @var Rule $rule */
+        $rule = $observer->getRule();
+        if ($rule->getIsActive()) {
+            return;
+        }
+        $this->iterateQuotes($rule);
+    }
+
+    /**
+     * To prevent out of memory issue we read row by row.
+     *
+     * @param Rule $rule
+     * @return void
+     */
+    private function iterateQuotes(Rule $rule): void
+    {
+        /** @var Collection $quoteCollection */
+        $quoteCollection = $this->collectionFactory->create();
+        $quoteCollection->getSelect()
+            ->where('is_active = 1')
+            ->where('FIND_IN_SET(?, applied_rule_ids)', $rule->getId());
+        $this->iterator->walk($quoteCollection->getSelect(), [[$this, 'callbackRecalculateQuote']]);
+    }
+
+    /**
+     * Callback to get and recalculate quote
+     *
+     * @param $args
+     * @return void
+     */
+    public function callbackRecalculateQuote($args): void
+    {
+        try {
+            /** @var Quote $quote */
+            $quote = $this->cartRepository->get($args['row']['entity_id']);
+            $quote->setAppliedRuleIds('');
+            $quote->collectTotals()->save();
```

Comment:
Hi. I will just add my 5 cents here. That's an interesting implementation. However, if we imagine a case with a large result set of quotes with the applied rule ID, this iterations loop might be really heavy for the system since collecting totals and saving the quote are really heavyweight operations. If we need to perform this action for 100-200 active quotes, we might end up with gateway timeout (best scenario).
Ideally, such actions should be delegated to the message queue and processed asynchronously. 
Just thinking out loud.

Reply:
Make sense. I thought that there is no rabbitMQ in "open source" edition... Is it good practice at all, as many users may not use ampq?

Reply:
I would suggest skipping totals calculation here at all. Just update quote to remove outdated rule.

Totals collection will be triggered on storefront when the user reloads his cart or performs any action on it.

After we verify that this works as expected, we may put message to queue to defer this action.

Reply:
There is the field trigger_recollect which I believe we could use for this.

Reply:
PR updated, tested with the condition ```'entity_id IN(?)' => range(0, 1000000)```